### ISSUE TITLE
SignalAwaiter acts weirdly if same signal is emited synchronously

### ISSUE THREAD

**Issue by MihaMarkic** [godotengine/godot#84254]
### Godot version

v4.1.2.stable.mono.official [399c9dc39]

### System information

Godot v4.1.2.stable.mono - Windows 10.0.19045 - Vulkan (Mobile) - dedicated NVIDIA GeForce RTX 2060 SUPER (NVIDIA; 31.0.15.3623) - AMD Ryzen 7 3700X 8-Core Processor (16 Threads)

### Issue description

Consider code below with two buttons. Click Start button and then Fire button.

I would expect to see these lines in Output panel:
```
Start button clicked
Waiting
Awaited
Re-emitted
```

However, it prints this:
```
Start button clicked
Waiting
Awaited
**Waiting**
Re-emitted
```

Seems like SignalAwaiter callback is called once more even though it has finished waiting and weirdly invokes WaitForAsync once more. Also later an exception is likely to happen.
If I include `await Task.Yield()`, then it works as expected.

```cs
public partial class root : Node2D
{
	[Signal]
	public delegate void TestEventHandler();
	
	void _on_start_button_pressed()
	{
		GD.Print("Start button clicked");
		await WaitForAsync();
	}
	
	async Task WaitForAsync()
	{
		GD.Print("Waiting");
		await ToSignal(this, SignalName.Test);
		GD.Print("Awaited");
		//await Task.Yield();
		EmitSignal(SignalName.Test);
		GD.Print("Re-emitted");
	}
	
	void _on_fire_button_pressed()
	{
		EmitSignal(SignalName.Test);
	}
}
```

### Steps to reproduce

Run attached sample, first click Start button and then Fire.

### Minimal reproduction project

[tosignalpossiblebug.zip](https://github.com/godotengine/godot/files/13218654/tosignalpossiblebug.zip)

**Comment by zaevi** [comment]
It also occurs error:
```
E 0:00:02:0307   NativeCalls.cs:6254 @ int Godot.NativeCalls.godot_icall_2_699(nint, nint, Godot.NativeInterop.godot_string_name, Godot.Variant[], Godot.NativeInterop.godot_string_name): Attempt to disconnect a nonexistent connection from '<Node2D#25685918925>'. Signal: 'Test', callable: 'Node2D(root.cs)::SignalAwaiterMiddleman::Test'.
  <C++ 错误>       Condition "!s->slot_map.has(*p_callable.get_base_comparator())" is true. Returning: false
  <C++ 源文件>      core\object\object.cpp:1419 @ Object::_disconnect()
  <栈追踪>          NativeCalls.cs:6254 @ int Godot.NativeCalls.godot_icall_2_699(nint, nint, Godot.NativeInterop.godot_string_name, Godot.Variant[], Godot.NativeInterop.godot_string_name)
                 GodotObject.cs:602 @ Godot.Error Godot.GodotObject.EmitSignal(Godot.StringName, Godot.Variant[])
                 root.cs:28 @ void root._on_fire_button_pressed()
                 root_ScriptMethods.generated.cs:44 @ bool root.InvokeGodotClassMethod(Godot.NativeInterop.godot_string_name&, Godot.NativeInterop.NativeVariantPtrArgs, Godot.NativeInterop.godot_variant&)
                 CSharpInstanceBridge.cs:24 @ Godot.NativeInterop.godot_bool Godot.Bridge.CSharpInstanceBridge.Call(nint, Godot.NativeInterop.godot_string_name*, Godot.NativeInterop.godot_variant**, int, Godot.NativeInterop.godot_variant_call_error*, Godot.NativeInterop.godot_variant*)

```

`ToSignal` gives you an one-shot awaiter, but this one is emitted twice before disconnecting.

So it can be also reproduced in GDScript:

```GDScript
extends Node

@export var btn : Button

func _ready():
	await btn.pressed
	btn.pressed.emit()

```

**Comment by MikeSchulze** [comment]
What I see on my tests the signal list is cleaned after some iteration. This results into error when awaiting on a signal.
I found a crazy workaround to hold the list by just calling `GetSignalConnectionList(<SignalName>)` 

It looks like the signal list is disposed for some reson.

**Comment by MikeSchulze** [comment]
Any update here, this is a major bug!
The signal connection list is cleaned for some reason, my workaround is not working well and runs randomly again into this error.

**Comment by AThousandShips** [comment]
Can you try on latest master? There's been improvements to signal management recently

**Comment by MikeSchulze** [comment]
> Can you try on latest master? There's been improvements to signal management recently

If there is a build I can download?

**Comment by AThousandShips** [comment]
You can download the latest artifacts, but mono is only available on Linux builds:
* https://github.com/godotengine/godot/actions/runs/8471735177

**Comment by kleonc** [comment]
Fixed by #89451 (since v4.3.dev6.official [89850d553]).
